v2/comments.go

```diff
@@ -69,15 +72,219 @@ func ExtractCommentTags(marker string, lines []string) map[string][]string {
+	values := tags[key]
 	if values == nil {
 		return defaultVal, nil
 	}
-	if values[0] == "true" {
+	if values[0].Value == "true" {
 		return true, nil
 	}
-	if values[0] == "false" {
+	if values[0].Value == "false" {
 		return false, nil
 	}
 	return false, fmt.Errorf("tag value for %q is not boolean: %q", key, values[0])
 }
+
+// ExtractFunctionStyleCommentTags parses comments for special metadata tags. The
+// marger argument should be unique enough to identify the tags needed, and
+// should not be a marker for tags you don't want, or else the caller takes
+// responsibility for making that distinction.
+//
+// The tagNames argument is a list of specific tags being extracted. If this is
+// nil or empty, all lines which match the marker are considered.  If this is
+// specified, only lines with begin with marker + one of the tags will be
+// considered.  This is useful when a common marker is used which may match
+// lines which fail this syntax (e.g. which predate this definition).
+//
+// This function looks for input lines of the following forms:
+//   - 'marker' + "key=value"
+//   - 'marker' + "key()=value"
+//   - 'marker' + "key(arg)=value"
+//
+// The arg is optional.  If not specified (either as "key=value" or as
+// "key()=value"), the resulting Tag will have an empty Args list.
+//
+// The value is optional.  If not specified, the resulting Tag will have "" as
+// the value.
+//
+// Tag comment-lines may have a trailing end-of-line comment.
+//
+// The map returned here is keyed by the Tag's name without args.
+//
+// A tag can be specified more than one time and all values are returned.  If
+// the resulting map has an entry for a key, the value (a slice) is guaranteed
+// to have at least 1 element.
+//
+// Example: if you pass "+" as the marker, and the following lines are in
+// the comments:
+//
+//	+foo=val1  // foo
+//	+bar
+//	+foo=val2  // also foo
+//	+baz="qux"
+//	+foo(arg)  // still foo
+//
+// Then this function will return:
+//
+//		map[string][]Tag{
+//	 	"foo": []Tag{{
+//				Name: "foo",
+//				Args: nil,
+//				Value: "val1",
+//			}, {
+//				Name: "foo",
+//				Args: nil,
+//				Value: "val2",
+//			}, {
+//				Name: "foo",
+//				Args: []string{"arg"},
+//				Value: "",
+//			}, {
+//				Name: "bar",
+//				Args: nil,
+//				Value: ""
+//			}, {
+//				Name: "baz",
+//				Args: nil,
+//				Value: "\"qux\""
+//		}}
+//
+// This function should be preferred instead of ExtractCommentTags.
+func ExtractFunctionStyleCommentTags(marker string, tagNames []string, lines []string) (map[string][]Tag, error) {
+	stripTrailingComment := func(in string) string {
+		parts := strings.SplitN(in, "//", 2)
+		return strings.TrimSpace(parts[0])
+	}
+
+	out := map[string][]Tag{}
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if len(line) == 0 {
+			continue
+		}
+		if !strings.HasPrefix(line, marker) {
+			continue
+		}
+		line = stripTrailingComment(line)
+		kv := strings.SplitN(line[len(marker):], "=", 2)
```

Comment:
How about using regex `^([a-zA-Z0-9-]+)(\(([^)][a-zA-Z0-9,]*)\))?(=(.*))?$`

tagKey=group[1]
args=group[3]
value=group[5]

Reply:
Hey everyone, I found the perl programmer!  :)

I love regexes, but only if I am the one writing them.  I am joking, but only sort of.

How confident are you that this is exactly correct?  How confident are you that it can handle future args?  There's a reason BNF and similar grammar-grammars exist.  I wanted to produce useful errors if it fails, and passing the `tagNames` down allows us to skip tag parsing entirely when the tag name doesn't match.  This is needed to handle older tags like:

```
+theDesiredTag(arg)=value
+protobuf.options.(gogoproto.goproto_stringer)=false
```

The second one would fail parsing but I don't want it to -- it's not what I am seeking.  The alternative is to go back to longer markers like `+k8s:validation` or `+k8s:` (and change a lot of internal logic.

The CEL rule parsing in kube-openapi uses `[0]` style indexing, and we may need to support that, eventually too.